src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyDispatcher.cs

```diff
@@ -0,0 +1,135 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Rendering;
+
+// When Blazor is deployed with multi-threaded runtime, WebAssemblyDispatcher will help to dispatch all JS interop calls to the main thread.
+// This is necessary because all JS objects have thread affinity. They are only available on the thread (WebWorker) which created them.
+// Also DOM is only available on the main (browser) thread.
+// The calls to InvokeAsync methods are dispatched synchronously and the returned Task is only resolved after the main thread finished the callback.
+internal sealed class WebAssemblyDispatcher : Dispatcher
+{
+    public static readonly Dispatcher Instance = new WebAssemblyDispatcher();
+    private readonly SynchronizationContext? _context;
+
+    private WebAssemblyDispatcher()
+    {
+        // capture the JSSynchronizationContext from the main thread.
+        _context = SynchronizationContext.Current;
+    }
+
+    public override bool CheckAccess() => SynchronizationContext.Current == _context || _context == null;
+
+    public override Task InvokeAsync(Action workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            workItem();
+            return Task.CompletedTask;
+        }
+
+        _context!.InvokeAsync(workItem);
+        return Task.CompletedTask;
```

Comment:
This looks like a mistake, unless I'm confused.

Shouldn't this return the task returned by `_context.InvokeAsync`, instead of `Task.CompletedTask`?

Reply:
It will invoke the `void SynchronizationContextExtension.InvokeAsync` and so there is no Task. Also it's blocking synchronous call.

Reply:
More details about the expected behavior in https://github.com/dotnet/aspnetcore/pull/48991/files#r1263731418